File: src/080RemoveDuplicatesFromSortedArrayII/080RemoveDuplicatesFromSortedArrayII.go
package main

import "fmt"

func removeDuplicates(nums []int) int {
	n := len(nums)
	if n <= 2 {
		return n
	}
	i := 0
	j := 1
	c := 1
	for j < n {
		if nums[i] == nums[j] {
			if c < 2 {
				nums[i+1], nums[j] = nums[j], nums[i+1]
				c++
				i++
			}
			j++
		} else {
			nums[i+1], nums[j] = nums[j], nums[i+1]
			i++
			j++
			c = 1
		}
	}
	return i + 1
}

func main() {
	nums := []int{1, 1, 1, 2, 2, 3}
	fmt.Printf("%v\n", nums[0:removeDuplicates(nums)])

	nums = []int{1, 2, 3, 2, 2, 2, 2, 3, 1, 3, 3}
	fmt.Printf("%v\n", nums[0:removeDuplicates(nums)])
}

File: src/080RemoveDuplicatesFromSortedArrayII/080RemoveDuplicatesFromSortedArrayII.rs
impl Solution {
    pub fn remove_duplicates(nums: &mut Vec<i32>) -> i32 {
        let (mut i, mut j, mut k, n) = (0, 0, 0, nums.len());
        const d: usize = 2;
        while j < n {
            if nums[j] != nums[i-k] {
                nums.swap(i, j);
                k = 1;
                i += 1;
            } else if nums[j] == nums[i-k] && k < d {
                nums.swap(i,j);
                k+=1;
                i+=1;
            }
            
            j += 1;
        }
        
        i as i32
    }
}
